src/SDK/MonitoringApplianceEmulator/Models/ObservableTask{TResult}.cs

```diff
@@ -0,0 +1,150 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="ObservableTask{TResult}.cs" company="Microsoft Corporation">
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// </copyright>
+//-----------------------------------------------------------------------
+
+namespace Microsoft.Azure.Monitoring.SmartDetectors.MonitoringApplianceEmulator.Models
+{
+    using System;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Occurs when the observable task was completed.
+    /// </summary>
+    /// <typeparam name="T">The type of the object that will be passed as an input parameter.</typeparam>
+    /// <param name="obj">The input parameter.</param>
+    public delegate void OnTaskCompletedEventHandler<in T>(T obj);
+
+    /// <summary>
+    /// An observable task, inherits from <see cref="ObservableObject"/>.
+    /// </summary>
+    /// <typeparam name="TResult">The type of the result being returned from the task. This type should have default constructor,
+    /// so it can be used as the task result as long as the task has not finished running</typeparam>
+    public class ObservableTask<TResult> : ObservableObject
+    {
+        private readonly ITracer tracer;
+
+        private Task<TResult> taskToRun;
+
+        private TResult result;
+
+        private bool isRunning;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ObservableTask{TResult}"/> class.
+        /// </summary>
+        /// <param name="taskToRun">The task that should be observed.</param>
+        /// <param name="tracer">The tracer.</param>
+        /// <param name="onTaskCompletedCallbabk">Method that should be executed after the task was completed.</param>
+        public ObservableTask(Task<TResult> taskToRun, ITracer tracer, OnTaskCompletedEventHandler<TResult> onTaskCompletedCallbabk = null)
+        {
+            this.TaskToRun = taskToRun;
+            this.tracer = tracer;
+
+            if (onTaskCompletedCallbabk != null)
+            {
+                this.OnTaskCompleted += onTaskCompletedCallbabk;
+            }
+
+            this.Result = default(TResult);
+            this.IsRunning = taskToRun.Status == TaskStatus.Running;
+
+            if (!taskToRun.IsCompleted)
+            {
+#pragma warning disable 4014
+                this.RunTaskAsync();
+#pragma warning restore 4014
+            }
+            else
+            {
+                this.Result = this.GetTaskResult();
+                this.OnTaskCompleted?.Invoke(this.Result);
+            }
```

Comment:
Why do you need this "if"? Can't you just call RunTaskAsync?

Reply:
this is not really new code, just moving of an existing code - rather not touch

---
In reply to: [231003699](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/13#discussion_r231003699) [](ancestors = 231003699)